### ISSUE TITLE
Improve test coverage

### ISSUE THREAD

**Issue by nicowilliams** [stedolan/jq#835]
Things to do:

 - Improve testing of `main.c`.  E.g., test colorized output (even if we don't test that it works functionally, we should run the code through valgrind).  We could strip out the escape sequences and check that colorized output parses equal to the input.

 - Improve testing of sytactic errors.
 - Improve testing of module system.
 - Improve testing of number parsing and formatting.
 - Improve testing of Unicode support.

**Comment by pkoppstein** [comment]
It would be very helpful to have an option to use --run-tests in a "quiet" mode so that one never saw any of the disassembly or trace messages.  This would, I think, lessen the disincentives to use --run-tests directly (e.g. in developing additional tests or on private test suites).

**Comment by nicowilliams** [comment]
Yeah, when tests fail usually I want to start with the Testing and ***
lines, and rarely do i want to look at jq bytecode disassembly.

**Comment by pkoppstein** [comment]
@nicowilliams wrote:
> EDIT: Done.

Excellent.

One last small (or at least smallish) enhancement for --run-tests would be to eliminate (or provide a way to eliminate) the automatic "self-test":
```
jq --run-tests all.test
[42,[42]]
{"foo":420,"bar":240}
jq: error: syntax error, unexpected INVALID_CHARACTER, expecting $end (Unix shell quoting issues?) at <top-level>, line 1:
}
jq: 1 compile error
...
```
The main reason for this request is the article at http://rosettacode.org/wiki/Test_a_function#jq

**Comment by nicowilliams** [comment]
Done.

**Comment by nicowilliams** [comment]
And since that page exists and `--run-tests` has been pretty stable, I documented it.

**Comment by pkoppstein** [comment]
The documentation for --run-tests contains a typo (consts) and
at least one now-inaccurate statement (about "%%FAIL").  I believe
the following is more accurate and perhaps slightly more
useful:
```   
   * `--run-tests [filename]`:

        Runs the tests in the given file or standard input.  This must
        be the last option given and does not honor all preceding
        options.  The input should consist of test cases, possibly
        interspersed with comment lines and empty lines; each test
        case consists of one program line followed in turn by one
        input line, as many lines of output as are expected, and a
        terminating empty line.  Each test case is run in isolation
        and results in at least one corresponding line of output.  If
        the actual and expected output differ, a line beginning with
        *** is emitted. Finally, a summary line such as "0 of 2 tests
        passed (1 malformed)" is also produced.

        Be warned that the behavior of this option is subject to change.
```
